axum-extra/src/either.rs

```diff
@@ -0,0 +1,233 @@
+//! `Either*` types for combining extractors or responses into a single type.
+//!
+//! # As an extractor
+//!
+//! ```
+//! use axum_extra::either::Either3;
+//! use axum::{body::Bytes, Json};
+//!
+//! async fn handler(
+//!     body: Either3<Json<serde_json::Value>, String, Bytes>,
+//! ) {
+//!     match body {
+//!         Either3::E1(json) => { /* ... */ }
+//!         Either3::E2(string) => { /* ... */ }
+//!         Either3::E3(bytes) => { /* ... */ }
+//!     }
+//! }
+//! #
+//! # let _: axum::Router = axum::Router::new().route("/", axum::routing::get(handler));
+//! ```
+//!
+//! Note that if all the inner extractors rejects the request then the rejection from the last
+//! extractor will be returned. For the example above that would be [`BytesRejection`].
+//!
+//! # As a response
+//!
+//! ```
+//! use axum_extra::either::Either3;
+//! use axum::{Json, http::StatusCode, response::IntoResponse};
+//! use serde_json::{Value, json};
+//!
+//! async fn handler() -> Either3<Json<Value>, &'static str, StatusCode> {
+//!     if something() {
+//!         Either3::E1(Json(json!({ "data": "..." })))
+//!     } else if something_else() {
+//!         Either3::E2("foobar")
+//!     } else {
+//!         Either3::E3(StatusCode::NOT_FOUND)
+//!     }
+//! }
+//!
+//! fn something() -> bool {
+//!     // ...
+//!     # false
+//! }
+//!
+//! fn something_else() -> bool {
+//!     // ...
+//!     # false
+//! }
+//! #
+//! # let _: axum::Router = axum::Router::new().route("/", axum::routing::get(handler));
+//! ```
+//!
+//! The general recommendation is to use [`IntoResponse::into_response`] to return different response
+//! types, but if you need to preserve the extract type then `Either*` works as well.
+//!
+//! [`BytesRejection`]: axum::extract::rejection::BytesRejection
+//! [`IntoResponse::into_response`]: https://docs.rs/axum/latest/axum/response/index.html#returning-different-response-types
+
+use axum::{
+    async_trait,
+    extract::{FromRequest, RequestParts},
+    response::{IntoResponse, Response},
+};
+
+/// Combines two extractors or responses into a single type.
+///
+/// See the [module docs](self) for examples.
+#[derive(Debug, Clone)]
+pub enum Either2<E1, E2> {
```

Comment:
This type is usually just called `Either`. But I don't care too much whether we have the `2` or not.